# Import required libraries

In [183]:
import pandas as pd
import numpy as np
from datetime import datetime


# Load the datasets

In [184]:
salesData_2017 = pd.read_csv(r"Sales Data/Sales Transactions-2017.csv")
salesData_2018 = pd.read_csv(r"Sales Data/Sales Transactions-2018.csv")
salesData_2019 = pd.read_csv(r"Sales Data/Sales Transactions-2019.csv")

In [185]:
# combine all the data into one
bronzeSalesData = pd.concat([salesData_2017,salesData_2018,salesData_2019])
bronzeSalesData.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN


# Explore data for initial analysis

In [186]:
print(f"Dimensions: {bronzeSalesData.shape}")
print(f"Column Datatypes: \n{bronzeSalesData.dtypes}")
print(f"Columns:\t{bronzeSalesData.columns}")

Dimensions: (111206, 9)
Column Datatypes: 
Date              str
Voucher           str
Party             str
Product           str
Qty               str
Rate              str
Gross             str
Disc              str
Voucher Amount    str
dtype: object
Columns:	Index(['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate', 'Gross', 'Disc',
       'Voucher Amount'],
      dtype='str')


In [187]:
print(f"Null data info:\n{bronzeSalesData.isnull().sum()}")
print(f"Statistical analysis on bronze data\n")
bronzeSalesData.describe(include='all')

Null data info:
Date               12591
Voucher            12557
Party                 40
Product            12591
Qty                12557
Rate               12558
Gross              12558
Disc              105609
Voucher Amount     83646
dtype: int64
Statistical analysis on bronze data



,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
count,98615,98649,111166,98615,98649,98648,98648,5597,27560
unique,836,10044,1994,867,546,1109,2548,371,7149
top,,,TP13,,100,,,,
freq,3053,3053,13056,3053,12528,3051,3053,3053,3053


#### Observations
The datatypes of the columns need to cleared for proper statistical information
- Date - to be updated as datetime
- Qty - to be updated as integer
- Rate, Gross, Disc, Voucher Amount - to be updated as float
- Disc - null can be updated as 0s

### Copy the raw data for initial cleaning

In [188]:
silverSalesData = bronzeSalesData.copy()

In [189]:
silverSalesData

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
19171,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
19172,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19173,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19174,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [190]:
silverSalesData = silverSalesData.replace(r"^\s*$", np.nan, regex=True) # to convert the empty string as proper NaN values
# convert all the numeric columns into numeric from string or object datatype
numeric_cols = [
    "Qty",
    "Rate",
    "Gross",
    "Disc",
    "Voucher Amount"
]

for col in numeric_cols:
    silverSalesData[col] = (
        silverSalesData[col]
        .astype("string")
        .str.replace(",", "", regex=False)
        .str.strip()
    )

    silverSalesData[col] = pd.to_numeric(
        silverSalesData[col],
        errors="coerce"
    )

In [191]:
# convert date into datetime datatype
silverSalesData["Date"] = pd.to_datetime(
    bronzeSalesData["Date"]
        .astype("string")
        .str.strip(),
    format="%d/%m/%Y",
    errors="coerce"
)
silverSalesData.dtypes

Date              datetime64[us]
Voucher                      str
Party                        str
Product                      str
Qty                      Float64
Rate                     Float64
Gross                    Float64
Disc                     Float64
Voucher Amount           Float64
dtype: object

In [192]:
silverSalesData.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,2017-04-01,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2.0,1690.0,3380.0,<NA>,13100.0
1,2017-04-01,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6.0,1620.0,9720.0,<NA>,<NA>
2,2017-04-01,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500.0,23.0,11500.0,<NA>,30990.0
3,2017-04-01,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6.0,1620.0,9720.0,<NA>,<NA>
4,2017-04-01,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5.0,1690.0,8450.0,<NA>,<NA>


In [193]:
for col in numeric_cols:
    source_not_empty = bronzeSalesData[col].notna()
    conversion_failed = source_not_empty & silverSalesData[col].isna()

    print(
        f"{col}: {conversion_failed.sum()} "
        "non-empty values failed numeric conversion"
    )

Qty: 3051 non-empty values failed numeric conversion
Rate: 3051 non-empty values failed numeric conversion
Gross: 3053 non-empty values failed numeric conversion
Disc: 3053 non-empty values failed numeric conversion
Voucher Amount: 3053 non-empty values failed numeric conversion


In [194]:
def normalize_source(series):
    return (
        series
        .astype("string")
        .str.strip()
        .replace("", pd.NA)
    )

In [195]:
conversion_failure_masks = {}

for col in numeric_cols:
    source_normalized = normalize_source(bronzeSalesData[col])

    failure_mask = (
        source_normalized.notna()
        & silverSalesData[col].isna()
    )

    conversion_failure_masks[col] = failure_mask

    print(
        f"{col}: {failure_mask.sum()} genuine conversion failures"
    )

Qty: 0 genuine conversion failures
Rate: 0 genuine conversion failures
Gross: 0 genuine conversion failures
Disc: 0 genuine conversion failures
Voucher Amount: 0 genuine conversion failures


In [196]:
print("Minimum date:", silverSalesData["Date"].min())
print("Maximum date:", silverSalesData["Date"].max())
print("Invalid dates:", silverSalesData["Date"].isna().sum())

Minimum date: 2017-04-01 00:00:00
Maximum date: 2019-10-10 00:00:00
Invalid dates: 15644


In [197]:
missingSummary = pd.DataFrame({
    "Missing Count": silverSalesData.isna().sum(),
    "Missing Percentage": (
        silverSalesData.isna().mean() * 100
    ).round(2)
})

missingSummary.sort_values(
    "Missing Percentage",
    ascending=False
)

,Missing Count,Missing Percentage
Disc,108662,97.71
Voucher Amount,86699,77.96
Date,15644,14.07
Product,15644,14.07
Voucher,15610,14.04
Gross,15611,14.04
Qty,15608,14.04
Rate,15609,14.04
Party,40,0.04


In [198]:
critical_transaction_cols = [
    "Date",
    "Voucher",
    "Party",
    "Product",
    "Qty",
    "Rate",
    "Gross"
]

transaction_mask = (
    silverSalesData[critical_transaction_cols]
    .notna()
    .all(axis=1)
)

In [199]:
print("Total rows:", len(silverSalesData))
print("Valid transaction rows:", transaction_mask.sum())
print("Other rows:", (~transaction_mask).sum())

Total rows: 111206
Valid transaction rows: 95561
Other rows: 15645


In [200]:
rejectedSalesData = (
    silverSalesData.loc[~transaction_mask]
    .copy()
)

validSalesData = (
    silverSalesData.loc[transaction_mask]
    .copy()
)
rejectedSalesData[
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
8,NaT,NaN,khader vali late en,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
9,NaT,NaN,try,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
11,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
12,NaT,NaN,DUMMY ENTRY,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
14,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
15,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
17,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
39,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
41,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
43,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>


In [201]:
rejectedSalesData.isna().sum()

Date              15644
Voucher           15610
Party                40
Product           15644
Qty               15608
Rate              15609
Gross             15611
Disc              15611
Voucher Amount    15611
dtype: int64

In [202]:
def get_missing_columns(row):
    return ", ".join(
        row.index[row.isna()].tolist()
    )

rejectedSalesData["Missing Columns"] = (
    rejectedSalesData.apply(
        get_missing_columns,
        axis=1
    )
)

rejectedSalesData[
    "Missing Columns"
].value_counts().head(20)

Missing Columns
Date, Voucher, Product, Qty, Rate, Gross, Disc, Voucher Amount           15602
Date, Party, Product                                                        34
Date, Voucher, Party, Product, Qty, Rate, Gross, Disc, Voucher Amount        6
Date, Voucher, Product, Gross, Disc, Voucher Amount                          2
Rate, Gross, Disc, Voucher Amount                                            1
Name: count, dtype: int64

## Distinguish comment rows from invalid transactions
A comment or continuation row typically has no date, voucher, product, quantity, rate, or gross.

In [203]:
comment_or_continuation_mask = (
    silverSalesData["Date"].isna()
    & silverSalesData["Voucher"].isna()
    & silverSalesData["Product"].isna()
    & silverSalesData["Qty"].isna()
    & silverSalesData["Rate"].isna()
    & silverSalesData["Gross"].isna()
)

A potentially invalid transaction has identifying information but is missing a critical numeric value:

In [204]:
invalid_transaction_mask = (
    silverSalesData["Date"].notna()
    & silverSalesData["Voucher"].notna()
    & silverSalesData["Product"].notna()
    & (
        silverSalesData["Qty"].isna()
        | silverSalesData["Rate"].isna()
        | silverSalesData["Gross"].isna()
    )
)

In [205]:
print(
    "Comments or continuation rows:",
    comment_or_continuation_mask.sum()
)

print(
    "Potential invalid transactions:",
    invalid_transaction_mask.sum()
)

Comments or continuation rows: 15608
Potential invalid transactions: 1


In [206]:
silverSalesData.loc[
    invalid_transaction_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
43740,2018-03-05,Sal:9266,CUSTOMER,TIRML-BL-KG-17*23,2.0,<NA>,<NA>,<NA>,<NA>


In [207]:
silverSalesData.loc[
    comment_or_continuation_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
8,NaT,NaN,khader vali late en,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
9,NaT,NaN,try,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
11,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
12,NaT,NaN,DUMMY ENTRY,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
14,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
15,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
17,NaT,NaN,dummy entry,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
39,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
41,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>
43,NaT,NaN,#NAME?,NaN,<NA>,<NA>,<NA>,<NA>,<NA>


### Create a row-status column
This provides an auditable cleaning process.

In [208]:
silverSalesData["Row Status"] = "Needs Review"

silverSalesData.loc[
    transaction_mask,
    "Row Status"
] = "Valid Transaction"

silverSalesData.loc[
    comment_or_continuation_mask,
    "Row Status"
] = "Comment or Continuation"

silverSalesData.loc[
    invalid_transaction_mask,
    "Row Status"
] = "Invalid Numeric Transaction"

In [209]:
silverSalesData["Row Status"].value_counts()

Row Status
Valid Transaction              95561
Comment or Continuation        15608
Needs Review                      36
Invalid Numeric Transaction        1
Name: count, dtype: int64

Then split the data:

In [210]:
rejectedSalesData = (
    silverSalesData[
        silverSalesData["Row Status"]
        != "Valid Transaction"
    ]
    .copy()
)

silverSalesData = (
    silverSalesData[
        silverSalesData["Row Status"]
        == "Valid Transaction"
    ]
    .copy()
)

silverSalesData.drop(
    columns=["Row Status"],
    inplace=True
)

In [211]:
silverSalesData.reset_index(
    drop=True,
    inplace=True
)

rejectedSalesData.reset_index(
    drop=True,
    inplace=True
)

In [212]:
# Fill missing discounts only in valid transaction rows
silverSalesData["Disc"] = (
    silverSalesData["Disc"].fillna(0)
)

print(
    "Missing discounts after filling:",
    silverSalesData["Disc"].isna().sum()
)

Missing discounts after filling: 0


### Clean text columns
Now standardize whitespace in the text fields.

In [213]:
text_cols = [
    "Voucher",
    "Party",
    "Product"
]

for col in text_cols:
    silverSalesData[col] = (
        silverSalesData[col]
        .astype("string")
        .str.strip()
        .str.replace(r"\s+", " ", regex=True)
    )

Preserve the original names while creating standardized versions for grouping:

Because these values might refer to the same party:
- Srinu
- SRINU
- srinu

In [214]:
silverSalesData["Party Standardized"] = (
    silverSalesData["Party"]
    .str.upper()
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

silverSalesData["Product Standardized"] = (
    silverSalesData["Product"]
    .str.upper()
    .str.strip()
    .str.replace(r"\s+", " ", regex=True)
)

## Analyze exact duplicate rows

In [215]:
exact_duplicate_count = (
    silverSalesData.duplicated().sum()
)

print(
    "Exact duplicate rows:",
    exact_duplicate_count
)

Exact duplicate rows: 275


In [216]:
duplicateRows = (
    silverSalesData[
        silverSalesData.duplicated(
            keep=False
        )
    ]
    .sort_values(
        by=[
            "Date",
            "Voucher",
            "Party",
            "Product"
        ]
    )
)

duplicateRows.head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount,Party Standardized,Product Standardized
1320,2017-04-12,Sal:439,TP13,Classic 250ml,100.0,23.0,2300.0,0.0,<NA>,TP13,CLASSIC 250ML
1335,2017-04-12,Sal:439,TP13,Classic 250ml,100.0,23.0,2300.0,0.0,<NA>,TP13,CLASSIC 250ML
1864,2017-04-17,Sal:443,SHAIK SIFULLA,CYCLE-BK-13*16,10.0,13.0,130.0,130.0,<NA>,SHAIK SIFULLA,CYCLE-BK-13*16
1873,2017-04-17,Sal:443,SHAIK SIFULLA,CYCLE-BK-13*16,10.0,13.0,130.0,130.0,<NA>,SHAIK SIFULLA,CYCLE-BK-13*16
1985,2017-04-17,Sal:552,TP13,Vishnu 300ml,100.0,27.5,2750.0,0.0,<NA>,TP13,VISHNU 300ML
1990,2017-04-17,Sal:552,TP13,Vishnu 300ml,100.0,27.5,2750.0,0.0,<NA>,TP13,VISHNU 300ML
2040,2017-04-18,Sal:497,Lakshmifancy,"TAPE ROLL 2"" 50MTRS",4.0,125.0,500.0,0.0,<NA>,LAKSHMIFANCY,"TAPE ROLL 2"" 50MTRS"
2041,2017-04-18,Sal:497,Lakshmifancy,"TAPE ROLL 2"" 50MTRS",4.0,125.0,500.0,0.0,<NA>,LAKSHMIFANCY,"TAPE ROLL 2"" 50MTRS"
2484,2017-04-21,Sal:599,SPM PROVISION-DAKKI,NO-00,1.0,150.0,150.0,0.0,<NA>,SPM PROVISION-DAKKI,NO-00
2485,2017-04-21,Sal:599,SPM PROVISION-DAKKI,NO-00,1.0,150.0,150.0,0.0,<NA>,SPM PROVISION-DAKKI,NO-00


In [217]:
# drop the duplicates
silverSalesData = (
    silverSalesData
    .drop_duplicates()
    .reset_index(drop=True)
)

## Validate Gross = Qty × Rate

In [218]:
silverSalesData["Calculated Gross"] = (
    silverSalesData["Qty"]
    * silverSalesData["Rate"]
)
silverSalesData["Gross Difference"] = (
    silverSalesData["Gross"]
    - silverSalesData["Calculated Gross"]
)
silverSalesData["Gross Valid"] = np.isclose(
    silverSalesData["Gross"],
    silverSalesData["Calculated Gross"],
    rtol=0,
    atol=0.01
)


In [219]:
silverSalesData[
    "Gross Valid"
].value_counts()

Gross Valid
True     95270
False       16
Name: count, dtype: int64

In [220]:
silverSalesData.loc[
    ~silverSalesData["Gross Valid"],
    [
        "Date",
        "Voucher",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Calculated Gross",
        "Gross Difference"
    ]
].head(20)

,Date,Voucher,Product,Qty,Rate,Gross,Calculated Gross,Gross Difference
716,2017-04-07,Sal:146,CYCLE-BLU-10*12,1600.0,5.42,8670.0,8672.0,-2.0
16633,2017-08-16,Sal:4219,"TAPEROLL 2""",12.0,83.33,1000.0,999.96,0.04
19558,2017-09-08,Sal:4921,KAVITHA GOLD 10*12,1400.0,5.57,7800.0,7798.0,2.0
19613,2017-09-09,Sal:4913,KAVITHA GOLD 10*12,700.0,5.57,3900.0,3899.0,1.0
26660,2017-11-14,Sal:6592,GARBAGE BAGS-KGS,30.32,120.0,3639.0,3638.4,0.6
30083,2017-12-16,Sal:7487,DURGA 13*16 PINK,1600.0,8.94,14310.0,14304.0,6.0
33386,2018-01-17,Sal:8221,BAHUBALI WINE,110.0,22.73,2500.0,2500.3,-0.3
36138,2018-02-19,Sal:8933,KRISHNA 10*12 RED,930.0,6.21,5775.0,5775.3,-0.3
39648,2018-03-26,Sal:9841,KRISHNA 10*12 RED,1200.0,6.23,7480.0,7476.0,4.0
39649,2018-03-26,Sal:9841,KRISHNA 13*16 RED,640.0,13.06,8360.0,8358.4,1.6


## Create basic validity flags

In [221]:
silverSalesData["Invalid Qty"] = (
    silverSalesData["Qty"] <= 0
)

silverSalesData["Invalid Rate"] = (
    silverSalesData["Rate"] < 0
)

silverSalesData["Invalid Gross"] = (
    silverSalesData["Gross"] < 0
)

silverSalesData["Invalid Discount"] = (
    silverSalesData["Disc"] < 0
)

In [222]:
validationSummary = pd.Series({
    "Invalid Qty": silverSalesData[
        "Invalid Qty"
    ].sum(),

    "Invalid Rate": silverSalesData[
        "Invalid Rate"
    ].sum(),

    "Invalid Gross": silverSalesData[
        "Invalid Gross"
    ].sum(),

    "Invalid Discount": silverSalesData[
        "Invalid Discount"
    ].sum(),

    "Gross Mismatch": (
        ~silverSalesData["Gross Valid"]
    ).sum()
})

validationSummary

Invalid Qty          0
Invalid Rate         0
Invalid Gross        0
Invalid Discount     0
Gross Mismatch      16
dtype: int64

In [223]:
# Missing discount means no recorded discount
silverSalesData["Disc"] = (
    silverSalesData["Disc"].fillna(0)
)

# Identify complete transaction rows
critical_transaction_cols = [
    "Date",
    "Voucher",
    "Party",
    "Product",
    "Qty",
    "Rate",
    "Gross"
]

transaction_mask = (
    silverSalesData[critical_transaction_cols]
    .notna()
    .all(axis=1)
)

print("Total rows:", len(silverSalesData))
print("Valid transaction rows:", transaction_mask.sum())
print("Rows requiring review:", (~transaction_mask).sum())

Total rows: 95286
Valid transaction rows: 95286
Rows requiring review: 0


In [224]:
silverSalesData.loc[
    ~transaction_mask,
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Disc",
        "Voucher Amount"
    ]
].head(50)

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount


## Inferences
- Main structural cleaning like numeric/date conversion, rejection of incomplete rows, text normalization, columns standardization, duplicate analysis/removal, gross validation
- The resulting dataset has 95,286 records with valid rows, while 16 gross mismatches, zero quantity etc., are present.
- The remaining business-level cleaning and recon needs to be done.

### Business-level Data Cleaning

Review the gross mismatches

In [225]:
grossMismatchData = (
    silverSalesData.loc[
        ~silverSalesData["Gross Valid"]
    ]
    .copy()
)

In [226]:
silverSalesData["Gross Difference Abs"] = (
    silverSalesData["Gross Difference"].abs()
)

silverSalesData["Gross Mismatch Type"] = np.select(
    [
        silverSalesData["Gross Valid"],
        silverSalesData["Gross Difference Abs"] <= 0.05,
        silverSalesData["Gross Difference Abs"] <= 1.00
    ],
    [
        "Exact / Within Tolerance",
        "Minor Rounding Difference",
        "Small Difference"
    ],
    default="Material Difference"
)

In [227]:
silverSalesData[
    "Gross Mismatch Type"
].value_counts()

Gross Mismatch Type
Exact / Within Tolerance     95270
Material Difference              8
Small Difference                 6
Minor Rounding Difference        2
Name: count, dtype: int64

In [228]:
silverSalesData.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount,Party Standardized,Product Standardized,Calculated Gross,Gross Difference,Gross Valid,Invalid Qty,Invalid Rate,Invalid Gross,Invalid Discount,Gross Difference Abs,Gross Mismatch Type
0,2017-04-01,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2.0,1690.0,3380.0,0.0,13100.0,SOLANKI PLASTICS,DONA-VAI-9100,3380.0,0.0,True,False,False,False,False,0.0,Exact / Within Tolerance
1,2017-04-01,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6.0,1620.0,9720.0,0.0,<NA>,SOLANKI PLASTICS,LITE FOAM(1200),9720.0,0.0,True,False,False,False,False,0.0,Exact / Within Tolerance
2,2017-04-01,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500.0,23.0,11500.0,0.0,30990.0,SARNESWARA TRADERS,VISHNU CHOTA WINE,11500.0,0.0,True,False,False,False,False,0.0,Exact / Within Tolerance
3,2017-04-01,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6.0,1620.0,9720.0,0.0,<NA>,SARNESWARA TRADERS,LITE FOAM(1200),9720.0,0.0,True,False,False,False,False,0.0,Exact / Within Tolerance
4,2017-04-01,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5.0,1690.0,8450.0,0.0,<NA>,SARNESWARA TRADERS,DONA-VAI-9100,8450.0,0.0,True,False,False,False,False,0.0,Exact / Within Tolerance


In [229]:
grossReviewData = (
    silverSalesData.loc[
        silverSalesData["Gross Mismatch Type"]
        != "Exact / Within Tolerance"
    ]
    .copy()
)

In [230]:
grossReviewData.shape

(16, 20)

In [231]:
grossReviewData = (
    grossReviewData
    .sort_values(
        "Gross Difference Abs",
        ascending=False
    )
)

grossReviewData[
    [
        "Date",
        "Voucher",
        "Party",
        "Product",
        "Qty",
        "Rate",
        "Gross",
        "Calculated Gross",
        "Gross Difference",
        "Gross Difference Abs",
        "Gross Mismatch Type"
    ]
]

,Date,Voucher,Party,Product,Qty,Rate,Gross,Calculated Gross,Gross Difference,Gross Difference Abs,Gross Mismatch Type
81722,2019-05-03,Sal:885,K.SRIHARI,LIDS-250ML,2000.0,0.01,10.0,20.0,-10.0,10.0,Material Difference
30083,2017-12-16,Sal:7487,SANKAR,DURGA 13*16 PINK,1600.0,8.94,14310.0,14304.0,6.0,6.0,Material Difference
57860,2018-09-05,Sal:4328,SVP-BUCHHI,MILAN 4*5 POLY,1050.0,12.26,12867.75,12873.0,-5.25,5.25,Material Difference
39648,2018-03-26,Sal:9841,KVK,KRISHNA 10*12 RED,1200.0,6.23,7480.0,7476.0,4.0,4.0,Material Difference
48141,2018-06-09,Sal:1908,SVP-BUCHHI,PAPER TEA,600.0,17.26,10353.0,10356.0,-3.0,3.0,Material Difference
716,2017-04-07,Sal:146,TEACHER,CYCLE-BLU-10*12,1600.0,5.42,8670.0,8672.0,-2.0,2.0,Material Difference
19558,2017-09-08,Sal:4921,BABASTORES,KAVITHA GOLD 10*12,1400.0,5.57,7800.0,7798.0,2.0,2.0,Material Difference
39649,2018-03-26,Sal:9841,KVK,KRISHNA 13*16 RED,640.0,13.06,8360.0,8358.4,1.6,1.6,Material Difference
19613,2017-09-09,Sal:4913,KHAN AND KHAN PLAST,KAVITHA GOLD 10*12,700.0,5.57,3900.0,3899.0,1.0,1.0,Small Difference
26660,2017-11-14,Sal:6592,KRISHNAPATNAM PORT,GARBAGE BAGS-KGS,30.32,120.0,3639.0,3638.4,0.6,0.6,Small Difference


Keeping all these records with flags for later review

In [232]:
# creatin overall data quality status
silverSalesData["Quality Status"] = np.where(
    silverSalesData["Gross Mismatch Type"]
    == "Exact / Within Tolerance",
    "Passed",
    "Review Gross Difference"
)

In [233]:
silverSalesData["Quality Status"].value_counts()

Quality Status
Passed                     95270
Review Gross Difference       16
Name: count, dtype: int64

In [234]:
# assign priority for review process
priority_mapping = {
    "Exact / Within Tolerance": "No Review",
    "Minor Rounding Difference": "Low",
    "Small Difference": "Medium",
    "Material Difference": "High"
}

silverSalesData["Review Priority"] = (
    silverSalesData[
        "Gross Mismatch Type"
    ]
    .map(priority_mapping)
)

In [235]:
silverSalesData[
    "Review Priority"
].value_counts()

Review Priority
No Review    95270
High             8
Medium           6
Low              2
Name: count, dtype: int64

Gross was validated against Qty × Rate using an absolute tolerance of 0.01. Of the 95,286 cleaned transaction rows, 95,270 were within tolerance. Sixteen records exceeded the tolerance, consisting of two minor rounding differences, six small differences, and eight material differences. These records were retained in the Silver dataset with quality and review-priority flags because the available data does not establish whether Gross, Rate, or another transaction rule is responsible for the variance.

#### Cleaning the voucher level reconcilation
- for that first let us create a meaningful transaction id

In [236]:
silverSalesData["Transaction Key"] = (
    silverSalesData["Date"].dt.strftime("%Y%m%d")
    + "_"
    + silverSalesData["Voucher"].astype("string")
    + "_"
    + silverSalesData["Party Standardized"]
)

In [237]:
print(
    "Unique transaction keys:",
    silverSalesData["Transaction Key"].nunique()
)

print(
    "Total transaction lines:",
    len(silverSalesData)
)

Unique transaction keys: 24474
Total transaction lines: 95286


Check whether one transaction key contains multiple party names:

In [238]:
transactionKeyCheck = (
    silverSalesData
    .groupby("Transaction Key")
    .agg(
        Party_Count=("Party Standardized", "nunique"),
        Date_Count=("Date", "nunique"),
        Voucher_Count=("Voucher", "nunique")
    )
)

transactionKeyCheck[
    (transactionKeyCheck["Party_Count"] > 1)
    | (transactionKeyCheck["Date_Count"] > 1)
    | (transactionKeyCheck["Voucher_Count"] > 1)
]

,Party_Count,Date_Count,Voucher_Count
Transaction Key,,,


#### Voucher summary

In [239]:
voucherSummary = (
    silverSalesData
    .groupby(
        [
            "Transaction Key",
            "Date",
            "Voucher",
            "Party Standardized"
        ],
        as_index=False
    )
    .agg(
        Line_Count=("Product", "size"),
        Product_Count=("Product Standardized", "nunique"),
        Quantity_Total=("Qty", "sum"),
        Gross_Total=("Gross", "sum"),
        Discount_Total=("Disc", "sum"),

        # Count how many rows contain a reported voucher amount
        Voucher_Amount_Count=("Voucher Amount", "count"),

        # Get the minimum and maximum to detect differing totals
        Voucher_Amount_Min=("Voucher Amount", "min"),
        Voucher_Amount_Max=("Voucher Amount", "max")
    )
)

#### Derive the reported voucher amount
- If one voucher total is recorded on only one transaction line, minimum and maximum will be the same.

In [240]:
voucherSummary["Reported Voucher Amount"] = (
    voucherSummary["Voucher_Amount_Max"]
)

In [241]:
voucherSummary["Calculated Net Amount"] = (
    voucherSummary["Gross_Total"]
    - voucherSummary["Discount_Total"]
)

In [242]:
voucherSummary["Voucher Difference"] = (
    voucherSummary["Reported Voucher Amount"]
    - voucherSummary["Calculated Net Amount"]
)

voucherSummary["Voucher Difference Abs"] = (
    voucherSummary["Voucher Difference"].abs()
)

#### Identify missing or inconsistent voucher totals

In [243]:
voucherSummary["Voucher Amount Missing"] = (
    voucherSummary["Reported Voucher Amount"].isna()
)
voucherSummary["Multiple Voucher Amount Entries"] = (
    voucherSummary["Voucher_Amount_Count"] > 1
)

In [244]:
voucherSummary["Conflicting Voucher Amounts"] = (
    voucherSummary["Voucher_Amount_Count"].gt(1)
    & ~np.isclose(
        voucherSummary["Voucher_Amount_Min"],
        voucherSummary["Voucher_Amount_Max"],
        rtol=0,
        atol=0.01,
        equal_nan=True
    )
)

#### Validate the voucher amount

In [245]:
voucherSummary["Voucher Valid"] = np.isclose(
    voucherSummary["Reported Voucher Amount"],
    voucherSummary["Calculated Net Amount"],
    rtol=0,
    atol=1.00,
    equal_nan=False
)

#### Classisifcation of the reconcilation results

In [246]:
voucherSummary["Voucher Reconciliation Status"] = np.select(
    [
        voucherSummary["Voucher Amount Missing"]
        .fillna(False)
        .to_numpy(dtype=bool),

        voucherSummary["Conflicting Voucher Amounts"]
        .fillna(False)
        .to_numpy(dtype=bool),

        voucherSummary["Voucher Valid"]
        .fillna(False)
        .to_numpy(dtype=bool),

        voucherSummary["Voucher Difference Abs"]
        .le(5.00)
        .fillna(False)
        .to_numpy(dtype=bool),

        voucherSummary["Voucher Difference Abs"]
        .le(20.00)
        .fillna(False)
        .to_numpy(dtype=bool)
    ],
    [
        "Missing Reported Voucher Amount",
        "Conflicting Reported Voucher Amounts",
        "Reconciled",
        "Small Difference",
        "Moderate Difference"
    ],
    default="Material Difference"
)

In [247]:
voucherSummary[
    "Voucher Reconciliation Status"
].value_counts(dropna=False)

Voucher Reconciliation Status
Reconciled                         10175
Material Difference                 8788
Moderate Difference                 5504
Small Difference                       6
Missing Reported Voucher Amount        1
Name: count, dtype: int64

In [248]:
print(
    "Missing classifications:",
    voucherSummary[
        "Voucher Reconciliation Status"
    ].isna().sum()
)

Missing classifications: 0


In [249]:
print(
    voucherSummary[
        "Voucher Difference Abs"
    ].le(5.00).dtype
)

print(
    small_difference_condition.dtype
)

boolean
bool


In [252]:
# create review priority of voucher recon
voucherPriorityMapping = {
    "Reconciled": "No Review",
    "Small Difference": "Low",
    "Moderate Difference": "Medium",
    "Material Difference": "High",
    "Missing Reported Voucher Amount": "High",
    "Conflicting Reported Voucher Amounts": "High"
}

voucherSummary["Review Priority"] = (
    voucherSummary[
        "Voucher Reconciliation Status"
    ]
    .map(voucherPriorityMapping)
)

In [253]:
voucherReconciliationSummary = (
    voucherSummary[
        "Voucher Reconciliation Status"
    ]
    .value_counts(dropna=False)
    .rename_axis("Reconciliation Status")
    .reset_index(name="Voucher Count")
)

voucherReconciliationSummary["Percentage"] = (
    voucherReconciliationSummary["Voucher Count"]
    / len(voucherSummary)
    * 100
).round(2)

voucherReconciliationSummary

,Reconciliation Status,Voucher Count,Percentage
0,Reconciled,10175,41.57
1,Material Difference,8788,35.91
2,Moderate Difference,5504,22.49
3,Small Difference,6,0.02
4,Missing Reported Voucher Amount,1,0.00


In [254]:
voucherSummary[
    "Review Priority"
].value_counts(dropna=False)

Review Priority
No Review    10175
High          8789
Medium        5504
Low              6
Name: count, dtype: int64

In [255]:
materialVoucherMismatchData = (
    voucherSummary.loc[
        voucherSummary[
            "Voucher Reconciliation Status"
        ] == "Material Difference"
    ]
    .sort_values(
        "Voucher Difference Abs",
        ascending=False
    )
)

materialVoucherMismatchData.head(30)

,Transaction Key,Date,Voucher,Party Standardized,Line_Count,Product_Count,Quantity_Total,Gross_Total,Discount_Total,Voucher_Amount_Count,...,Reported Voucher Amount,Calculated Net Amount,Voucher Difference,Voucher Difference Abs,Voucher Amount Missing,Multiple Voucher Amount Entries,Conflicting Voucher Amounts,Voucher Valid,Voucher Reconciliation Status,Review Priority
10656,20180426_Sal:648_SEKHAR MARKET,2018-04-26,Sal:648,SEKHAR MARKET,6,6,12490.0,29165.0,0.0,1,...,45725.0,29165.0,16560.0,16560.0,False,False,False,False,Material Difference,High
24148,20190927_Sal:4608_SEKHAR MARKET,2019-09-27,Sal:4608,SEKHAR MARKET,3,1,180.0,50400.0,0.0,1,...,66000.0,50400.0,15600.0,15600.0,False,False,False,False,Material Difference,High
19923,20190413_Sal:406_TP13,2019-04-13,Sal:406,TP13,32,32,4872.0,83461.0,0.0,1,...,97032.0,83461.0,13571.0,13571.0,False,False,False,False,Material Difference,High
1407,20170515_Sal:1506_TP13,2017-05-15,Sal:1506,TP13,54,47,8016.0,152007.0,0.0,1,...,163882.0,152007.0,11875.0,11875.0,False,False,False,False,Material Difference,High
20600,20190508_Sal:1049_TP13,2019-05-08,Sal:1049,TP13,13,12,881.0,32230.0,0.0,1,...,43625.0,32230.0,11395.0,11395.0,False,False,False,False,Material Difference,High
19808,20190409_Sal:263_TP13,2019-04-09,Sal:263,TP13,31,31,7934.0,93318.0,0.0,1,...,104144.0,93318.0,10826.0,10826.0,False,False,False,False,Material Difference,High
17109,20181226_Sal:7272_TP13,2018-12-26,Sal:7272,TP13,63,53,8429.0,173741.0,0.0,1,...,183461.0,173741.0,9720.0,9720.0,False,False,False,False,Material Difference,High
6067,20171023_Sal:6338_TP13,2017-10-23,Sal:6338,TP13,32,32,5804.0,84625.0,0.0,1,...,93975.0,84625.0,9350.0,9350.0,False,False,False,False,Material Difference,High
11920,20180609_Sal:2067_TP13,2018-06-09,Sal:2067,TP13,20,20,3876.0,63641.0,0.0,1,...,72761.0,63641.0,9120.0,9120.0,False,False,False,False,Material Difference,High
11767,20180604_Sal:2052_TP13,2018-06-04,Sal:2052,TP13,22,20,5003.0,75332.0,0.0,1,...,83782.0,75332.0,8450.0,8450.0,False,False,False,False,Material Difference,High


Inspect missing voucher amounts

In [256]:
missingVoucherAmountData = (
    voucherSummary.loc[
        voucherSummary[
            "Voucher Amount Missing"
        ]
    ]
    .copy()
)

missingVoucherAmountData[
    [
        "Date",
        "Voucher",
        "Party Standardized",
        "Line_Count",
        "Gross_Total",
        "Discount_Total",
        "Calculated Net Amount"
    ]
].head(30)

,Date,Voucher,Party Standardized,Line_Count,Gross_Total,Discount_Total,Calculated Net Amount
9166,2018-03-02,Sal:9204,SB.ENTERPRISES,1,12100.0,12100.0,0.0


In [257]:
voucherStatusColumns = [
    "Transaction Key",
    "Calculated Net Amount",
    "Reported Voucher Amount",
    "Voucher Difference",
    "Voucher Reconciliation Status",
    "Review Priority"
]

silverSalesData = silverSalesData.merge(
    voucherSummary[voucherStatusColumns],
    on="Transaction Key",
    how="left",
    validate="many_to_one"
)

In [258]:
print(
    "Silver rows after voucher merge:",
    len(silverSalesData)
)

Silver rows after voucher merge: 95286


Final quality checks

In [259]:
assert (
    voucherSummary["Transaction Key"].is_unique
), "Voucher summary contains duplicate transaction keys"

In [260]:
assert (
    silverSalesData[
        "Voucher Reconciliation Status"
    ].notna().all()
), "Some Silver rows did not receive a reconciliation status"

In [261]:
print(
    "Total vouchers:",
    len(voucherSummary)
)

print(
    "Reconciled vouchers:",
    voucherSummary["Voucher Valid"].sum()
)

print(
    "Vouchers requiring review:",
    (
        voucherSummary[
            "Voucher Reconciliation Status"
        ] != "Reconciled"
    ).sum()
)

Total vouchers: 24474
Reconciled vouchers: 10175
Vouchers requiring review: 14299
